### ISSUE TITLE
[RFE] Request to block delimited changes in /etc/hosts

### ISSUE THREAD

**Issue by nhojpatrick** [crc-org/crc#2992]
### General information

  * OS: macOS
  * Hypervisor: hyperkit
  * Did you run `crc setup` before starting it (Yes/No)? Yes
  * Running CRC on: Laptop

When `crc start` runs it always rewrites `/etc/hosts`, even if it doesn't need to change anything.

Things done when hosts file is rewritten;
1) comments removed
2) empty lines removed
3) whitespace squashed into single space

## CRC version
```bash
CodeReady Containers version: 1.39.0+ba8a8acc
OpenShift version: 4.9.15 (bundle installed at /Applications/CodeReady Containers.app/Contents/Resources/crc_hyperkit_4.9.15.crcbundle)
```
  
## CRC status
```bash
n/a
```

## CRC config
```bash
n/a
```

## Host Operating System
```bash
ProductName:	Mac OS X
ProductVersion:	10.15.7
BuildVersion:	19H1713
```

### Steps to reproduce

  1. edit `/etc/hosts`, add comment, add extra new line, add extra whitespace
  2. `crc start`

### Expected

If crc needs to update the hosts file, can it add it's own line or own section like Docker Desktop, e.g;
```bash
# Managed by Code-Ready CRC
127.0.0.1 api.crc.testing canary-openshift-ingress-canary.apps-crc.testing console-openshift-console.apps-crc.testing default-route-openshift-image-registry.apps-crc.testing downloads-openshift-console.apps-crc.testing oauth-openshift.apps-crc.testing
# End of section
```

### Actual

`crc start` always rewrite `/etc/hosts`.
- comments are lost
- whitespace is squashed to a space
- empty lines are list
- marker sections use by other apps are lost

### Logs

n/a

**Comment by cfergeau** [comment]
> I'd have expected this change to /etc/hosts to be done by 'crc setup', not 'crc start'. It seems to me, without digging into the code, that all the information is already known at 'crc setup', no? The IP address (hard-coded?) also appears in NetworkManager/dnsmasq.d/crc.conf, which is created by 'crc setup'.

The IP address is/was dynamic on some platforms. The main reason why it can't be done at `setup` time though is because `/etc/hosts` needs to be updated when you create a route for your own application (`myapp.crc-apps.testing`), we can't add wildcard hostnames in `/etc/hosts` for `*.apps-crc.testing`, so this needs to be updated on the fly.

**Comment by nhojpatrick** [comment]
> Hi,
> 
> Not sure what's the exact meaning of '[RFE] Request to block delimited changes in /etc/hosts' (not being a native English speaker, for that matter), but I also now encountered all of the issues mentioned above:
> 
> * comments removed - most annoying part, for me, as I had several such comments
> * empty lines removed - not that important, but still
> * whitespace squashed into single space - also not that important. Main issue with it is that it makes it hard to understand what 'crc start' actually meant to change in the file (by comparing it with a copy. I maintain /etc as a local git repo).
> 
> I'd also like to add my own item:
> 
> * I'd have expected this change to /etc/hosts to be done by 'crc setup', not 'crc start'. It seems to me, without digging into the code, that all the information is already known at 'crc setup', no? The IP address (hard-coded?) also appears in NetworkManager/dnsmasq.d/crc.conf, which is created by 'crc setup'.

Me neither, that wasn't my original title. It was "anti social behaviour regarding hosts file management"

**Comment by cfergeau** [comment]
> I totally understand you need to update the hosts file.
But why do it in such anti social way?

I was only answering to the part which I quoted, not about this issue in general ;) Not the first time this comes up though https://github.com/code-ready/crc/issues/2710#issuecomment-915990215
Main reason why this happens is because the go module we use for this (github.com/goodhosts/hostsfile) behaves this way. We could switch to something different, but this is low priority at the moment.
Maybe there are some easy improvements which could be made on our side though, as https://github.com/goodhosts//cli seems to be nicer in the changes it makes to the hosts file.

**Comment by nhojpatrick** [comment]
any update if this will be dealt with, i've multiple programs touch hosts and use comments to work out which section they added and can edit, but crc just drops all comments and rewrites it, which then breaks other applications which have spent time making sure they are a good collaborate program.
